More Source Generators

Redth opened the pull request:
This adds a few things around source generators... 

 - SourceGen.Util - contains helper extension methods and types for writing source generators, and can be shared by multiple projects
 - SourceGen.Core - for source generators not specific to MAUI Controls (more later)
 - TestUtils.SourceGen - some base types for source gen related test projects
 - Controls.SourceGen.Tests - test project for the controls generators (XamlG)
 - SourceGen.Core.Tests - test project for the core generators


# New Core Source Generators

## Font Configuration Source Generator

Generates an extension method for the app builder: `builder.AutoConfigureFonts(fonts => {});` which will automatically emit statements for each `MauiFont` item in the project (and its FontAlias) - the delegate can still be used to add/override the generated emits just the same as if you used `builder.ConfigureFonts(fonts => {});`

## App Builder Source Generator
This attempts to generate the boilerplate AppDelegate/Main for Mac/iOS and Application/MainActivity if it's not present in your code.

## static MauiApp method

The generator looks for a static method which returns a `MauiApp` instance.  This is assumed to be the main entry point in the app and what the generated classes will invoke.

## iOS
- Program/Main - if no `static void Main(string[] args)` type method is found, it will be generated to reference the AppDelegate
- AppDelegate - if no subclass of `MauiUIApplicationDelegate` is found, it will generate a partial class in the root namespace of the app along with the `override MauiApp CreateMauiApp()` If there already is a partial subclass, it will check if the `CreateMauiApp` is overridden and generate it if not.

## Android
- MainActivity - if no Activity type is found that subclasses `MauiAppCompatActivity` and is marked with `MainLauncher=true`, it will generate a partial class for it.
- Application - if no subclass of `MauiApplication` or `Application` is found, it will create a partial one along with the `override MauiApp CreateMauiApp()` If there already is a partial subclass, it will check if the `CreateMauiApp` is overridden and generate it if not.

Redth committed:
Add sourcegen tests, font sourcegen, app boilerplate sourcegen

Redth committed:
Merge branch 'main' into more-source-gen

Redth committed:
-net6 isn't needed in project name here

Redth committed:
Fix path to proj rename

Redth committed:
Add project for shared sourcegen util classes/extension methods

Redth committed:
Add project for shared sourcegen test project things

Redth committed:
Add core source generator project

Redth committed:
Move out files that went to other projects

Redth committed:
Use helper methods in XamlG source generator

Redth committed:
Add a basic test for xamlg

Redth committed:
Refactor builder generator into platforms

Redth committed:
Split out app builder tests into ios/android

Redth committed:
Fix bug where type part info missed last type name part

Redth committed:
Merge branch 'main' into more-source-gen

Redth committed:
Fix refs to renamed proj

Redth committed:
Remove line from bad merge

Redth committed:
Rename gen types

Redth committed:
Add new sourcegen projects to sln/slnf/sdk projs

Redth committed:
Ensure source gen required properties/item group metadata are visible

mattleibow commented on src/Controls/tests/SourceGen.Tests/TestData/SimpleTest.xaml:
```diff
@@ -0,0 +1,12 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
```
This is the old forms namespace.

mattleibow commented on src/Controls/tests/SourceGen.Tests/Controls.SourceGen.UnitTests.csproj:
```diff
@@ -0,0 +1,38 @@
+    <ProjectReference Include="..\..\..\TestUtils\src\TestUtils.SourceGen\TestUtils.SourceGen.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <None Update="TestData\SimpleTest.xaml">
+      <Generator></Generator>
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
```
IDE happy times?

Redth replied:
Maybe just the blank generator attribute, but the point is to have this file available to test so copying it to the expected location seems fine?

mattleibow commented on src/SourceGen/src/Core/AppBuilderGenerator.cs:
```diff
@@ -0,0 +1,365 @@
+				if (!HasMainLauncher && syntaxNode is AttributeSyntax attrSyntax)
+				{
+					var attrFullName = attrSyntax.Name.ToFullString();
+					if (attrSyntax.Name.ToFullString() == "global::Android.App.Activity")
```
Activity or ActivityAttribute?

mattleibow commented on src/SourceGen/src/Core/SourceGen.Core.csproj:
```diff
@@ -0,0 +1,56 @@
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.9.0" PrivateAssets="All" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <!-- MauiFont items -->
+    <CompilerVisibleItemMetadata Include="AdditionalFiles" MetadataName="SourceItemGroup" Visible="false" />
+    <CompilerVisibleItemMetadata Include="AdditionalFiles" MetadataName="FontAlias" Visible="false" />
+
+    <!-- Properties for emitting default entry boilerplate -->
+    <CompilerVisibleProperty Include="OutputType" />
+    <CompilerVisibleProperty Include="TargetPlatformIdentifier" />
+    <CompilerVisibleProperty Include="ApplicationTitle" />
+  </ItemGroup>
```
Why does the source project need these properties/items?

pictos commented on src/SourceGen/src/Core/AppBuilderGenerator.cs:
```diff
@@ -0,0 +1,365 @@
+					{
+						var returnTypeFullNamme = context.GetReturnType(mds);
+
+						if (returnTypeFullNamme == "Microsoft.Maui.MauiApp")
```
maybe we can rely on symbols instead of strings, to make it safer (?)

mattleibow replied:
Yeah. I totally forgot my usual "this should at the very least be a const and not magic string littered about". 😄

pictos commented on src/SourceGen/src/Util/SyntaxNodeExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+			return info;
+		}
+
+		public static string GetNamespace(this TypeDeclarationSyntax source)
```
We can access the namespace using the symbol of the `Syntax`. The `Compilation` is from `GeneratorExecutionContext` parameter from `ISourceGenerator.Execute` method.

```csharp
 void GetNamespace(Compilation compilation, TypeDeclarationSyntax type)
        {
            var model = compilation.GetSemanticModel(type.SyntaxTree);
            var symbol = model.GetDeclaredSymbol(type);
            var typeNamespace = symbol.ContainingNamespace;
        }
```

pictos commented on src/SourceGen/src/Util/SyntaxNodeExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+			return result.ToString().TrimEnd('.');
+		}
+
+		public static string GetFullName(this TypeDeclarationSyntax source)
```
If the `GetFullName` is the full name of a type, like `namespace + Class + Member` you can access that using symbol

```csharp
void GetNamespace(Compilation compilation, TypeDeclarationSyntax type)
        {
            var model = compilation.GetSemanticModel(type.SyntaxTree);
            var symbol = model.GetDeclaredSymbol(type);
            var fullname = symbol.ToDisplayString();
        }
```

pictos reviewed (commented):
I have some suggestions for this code... Also, I would like to remember to add a `// <auto-generated/>` comment at the top of generated classes to make analyzers skip it, as I reported at #2301

StephaneDelcroix commented on src/SourceGen/src/Util/GeneratorSyntaxContextExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+			if (interfaceType.Symbol == null)
+				return null;
+			var s = GetFullName(interfaceType.Symbol);
+			if (s != null && !string.IsNullOrEmpty(s))
```
you're repeating yourself :)

StephaneDelcroix commented on src/SourceGen/src/Util/GeneratorSyntaxContextExtensions.cs:
```diff
@@ -0,0 +1,45 @@
+				return first;
+			if (string.IsNullOrWhiteSpace(first))
+				return second;
+			return $"{first}.{second}";
```
string.join ?

StephaneDelcroix commented on src/SourceGen/src/Util/SyntaxNodeExtensions.cs:
```diff
@@ -0,0 +1,144 @@
+		const char NESTED_CLASS_DELIMITER = '+';
+		const char NAMESPACE_CLASS_DELIMITER = '.';
+		const char TYPEPARAMETER_CLASS_DELIMITER = '`';
+
```
all the type declaration should be prefixed with `global::`, to avoid name clashes with user namespace

StephaneDelcroix replied:
but you're doing that later on

StephaneDelcroix reviewed (approved):
👍

Redth committed:
Split sourcegen props/targets for Core/Controls

Redth committed:
Add `// <auto-generated/>` to source gen'd code

Redth committed:
Fix test data xaml for maui

Redth committed:
Fix rename

Redth committed:
Clean up extensions

Redth committed:
Tests need to reference platform related Maui dll's

Otherwise types the generator is looking for specific to a platform cannot be found.

Redth committed:
Cleanup

Redth committed:
Use Symbols in source gen as much as possible

This should bring consistency by getting symbols with full type info for analysis, instead of checking partial strings which could be fully qualified, partially qualified, or not for instance.

Redth committed:
Use Try version of method to make Stephane happy

pictos commented on src/SourceGen/src/Core/AppBuilderGenerator.cs:
```diff
@@ -0,0 +1,406 @@
+				// Look for an AppDelegate implementation of some sort
+				if (syntaxNode is BaseTypeSyntax baseTypeSyntax)
+				{
+					var baseTypeSymbol = context.SemanticModel.Compilation.GetSemanticModel(syntaxNode.SyntaxTree)
```
you have the compilation at line 69
```suggestion
					var baseTypeSymbol = compilation.GetSemanticModel(syntaxNode.SyntaxTree)
```

pictos commented on src/SourceGen/src/Core/AppBuilderGenerator.cs:
```diff
@@ -0,0 +1,406 @@
+
+					var baseTypeName = baseTypeSymbol?.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat) ?? String.Empty;
+
+					if (baseTypeName.Equals(MauiUIApplicationDelegateFullyQualifiedName))
```
@Redth I believe that we can use symbols here. Are you open to PRs? If yes I can do something on my side and send a PR to your branch

Redth committed:
Merge branch 'main' into more-source-gen

Redth committed:
Use some fake assemblies to avoid mono android dll ref in tests

Redth committed:
Fix sln after merge

rmarinho commented:
@Redth do we still want to get this in?
